CONVENTION.md

```diff
@@ -0,0 +1,453 @@
+
+1. [ ] Stand fast - Stick with snake_case where you can, mixedCase where you must
+2. [ ] Give in - Use mixedCase everywhere
+3. [x] Compromise - Use whichever convention is the majority of code per module or package
+
+##### Stand Fast
+
+This is what I typically do and see.
+
+```py
+from Qt import QtWidgets
+
+class Window(QtWidgets.QDialog):
+	def __init__(self, parent=None):
+		super(Window, self).__init__(parent)
+
+		my_button = QtWidgets.QPushButton("Press me")
+		some_layout = QtWidgets.QHBoxLayout(self)
+		some_layout.addWidget(my_button)
+
+		my_button.clicked.connect(self.on_button_clicked)
+
+	def mousePressEvent(self, event):
+		pass
+
+	def on_button_clicked(self):
+		pass
+```
+
+Which is clearly horrendous. The few occasions where Qt appears to play ball - `clicked` and `connect` - are circumstantial. That isn't snake case, it's mixedCase masquerading as snake_case, ready to jump out at you at a moments notice.
+
+One of the advantages (or so I tell myself) is that it's obvious which methods are overridden, and which are new. But is that enough to justify the overhead in reading ability? It's like reading-an ENGlish `_sen_tenc_e_` with 2 deffirent ConvenTions. Takes you out of the story.
+
+#### Give in
+
+Why not just write everything in mixedCase and get it over with?
+
+If your code consists of only Qt and your own code, that's a fine idea; but standard and third-party libraries don't work that way.
+
+```py
+import pymongo
+
+class Model(QtWidgets.QTableModel):
+	def setData(self, index, value, role):
+		item = pymongo.find_one({"type": "asset"})
+		...
+```
+
+<br>
+
+#### Compromise
+
+Optimise to reduce context switching. Use the majority convention; less granularity is better.
+
+1. Per project
+1. Per package
+2. Per module
+3. Per class
+3. Per method
+3. Per block
+
+For example, if a package (e.g. `avalon.loader`) is mostly Qt, use mixedCase. But if it isn't clear (e.g. `avalon.maya`) use mixedCase for modules where mixedCase is a majority, snake_case otherwise. If it isn't clear *within* a module which is a majority, stay consistent within a class, method or worst case, per block of code.
+
+**Bad**
+
+Sometimes you have no choice but to be consistent per block, but don't do this.
+
+- Each line has its own convention
+- One line has two conventions
+- Context switch count: 4
+
+```py
+class Window(QtWidgets.QDialog):
+	def __init__(self):
+		some_button = QtWidgets.QPushButton("Hello")     # 1
+		myLabel = QtWidgets.QLabel("World")              # 2
+		this_layout = QtWidgets.QVBoxLayout(self)        # 3
+		that_widget = QtWidgets.QWidget()                #
+		otherLayout = QtWidgets.QHBoxLayout(that_widget) # 4
+```
+
+**Good**
+
+Instead do this.
+
+- Context switch count: 2
+
+```py
+class Window(QtWidgets.QDialog):
+	def __init__(self):
+		some_button = QtWidgets.QPushButton("Hello")    # 1
+		this_layout = QtWidgets.QVBoxLayout(self)       #
+
+		thatWidget = QtWidgets.QWidget()                # 2
```

Comment:
> This seems like a bad example. Why are we using this a "Good" example for a compromise? I think the compromises should be made solely where they cannot be avoided, whereas this one easily can.

What is a better example?

> I would opt actually for Avalon tools using snake_case as that is the code guidelines for Avalon itself, PEP08. 

For clarity, you are referring to "stand fast"? If not, how does your ideal case differ?

- https://github.com/mottosso/avalon-core/blob/convention-guidelines/CONVENTION.md#stand-fast

Reply:
Just I was referring to the *Stand Fast* code where all variables are PEP08 lowercase with underscores. I don't see a reason to mix the variable naming conventions inside of Avalon, if it's a "sometimes" thing it'll only make room for discussion/debating that should just have gotten a clear cut rule.

Reply:
> Just I was referring to the Stand Fast code where all variables are PEP08 lowercase with underscores. I don't see a reason to mix the variable naming conventions inside of Avalon, if it's a "sometimes" thing it'll only make room for discussion/debating that should just have gotten a clear cut rule.

But it's not just variables; it's methods too, and supporting functions. The example highlights that (1) if we want snake_case, we must still have mixedCase and (2) that's ugly.

Consistent use of ugly, non-standard mixedCase is greater than inconsistent use of a nice and standard use of snake_case.

But I can see how we'd get the question "what should I use where?", as "majority of code" changes over time. It could start out with mostly mixedCase and turn into mostly snake_case. Do we then refactor to one over the other? The guideline could be "anything widgets use mixedCase, everything else use snake_case".

Not 100% sure, anyone else have any thoughts here?

Reply:
> But it's not just variables; it's methods too, and supporting functions. 

Is it? It's really just methods and functions, right? Or at least, the things you reference from other libraries? What I'm saying is that wherever you don't **require** to match exactly the mixedCase or any other standard, then we don't.

So yes, we do override `mousePressEvent` exactly as it is, but otherwise it doesn't work. However, we do define the callback methods as `on_accept` or alike.

The rule then would be to use `snake_case` unless the code wouldn't work when using `snake_case`.

Reply:
I think we're saying the same thing.

If we use snake_case where we can, mixedCase elsewhere; that's inconsistent. That's what this bullet point addresses. The question is whether to do that, or to do something about it. What to do? That's what the three options are for; stand fast, give in, or compromise.

Reply:
So the default is *stand fast*, and other options are just optional ?
I actually lean to all *stand fast* as default, since it’s most simple.

Reply:
Yes, then I agree. It's stand fast that is the **Good** example. Also, the three examples and only one being the right one didn't stand out to me. How about just identifying them as **Good**, **Bad** and **Bad** similar to how you did with the others.